libkatana_python_native/src/Reductions.cpp

```diff
@@ -0,0 +1,90 @@
+#include <katana/python/Conventions.h>
+#include <katana/python/NumbaSupport.h>
+#include <katana/python/PythonModuleInitializers.h>
+#include <katana/python/TemplateSupport.h>
+#include <pybind11/pybind11.h>
+
+namespace py = pybind11;
+
+namespace {
+
+// TODO(amp): Use template template parameters once we can rely on having them.
+struct ForGAccumulator {
+  template <typename T>
+  using type = katana::GAccumulator<T>;
+};
+
+struct ForReduceMin {
+  template <typename T>
+  using type = katana::GReduceMin<T>;
+};
+
+struct ForReduceMax {
+  template <typename T>
+  using type = katana::GReduceMax<T>;
+};
+
+struct ForReduceLogicalOr {
+  template <typename T>
+  using type = katana::GReduceLogicalOr;
+};
+
+struct ForReduceLogicalAnd {
+  template <typename T>
+  using type = katana::GReduceLogicalAnd;
+};
+
+template <typename For>
+struct ReducibleFunctor {
+  template <typename T>
+  py::object operator()(py::module_& m, const char* name) {
+    py::class_<typename For::template type<T>> cls(
+        m, name,
+        "A reducer object that can updated with new values and combines the "
+        "results efficiently using the appropriate operator.\n"
+        "\n"
+        "This class can be passed into numba compiled code and it's methods "
+        "can be used from there.\n");
+    cls.def(py::init<>());
+    cls.def(
+        py::init([](T v) {
+          auto self = std::make_unique<typename For::template type<T>>();
+          self->update(v);
+          return self;
+        }),
+        "Create a new instance with an initial value.");
+    katana::RegisterNumbaClass(cls);
+    katana::DefWithNumba<const T&>(
+        cls, "update", &For::template type<T>::update,
+        "Update this reducer with ``v`` performing the operation.");
+    katana::DefWithNumba<>(
+        cls, "reduce", &For::template type<T>::reduce,
+        "Get the current value of the reducer. This must only be called from "
+        "single threaded code.");
+    katana::DefWithNumba<>(
+        cls, "get_local", &For::template type<T>::getLocal,
+        "Get a sub-result of the reducers operation. This is generally the "
+        "reduced value for this thread.");
+    katana::DefWithNumba<>(
+        cls, "reset", &For::template type<T>::reset,
+        "Reset the reducer to its zero. This must only be called from single "
+        "threaded code.");
+    katana::DefConventions(cls);
+    return std::move(cls);
+  }
+};
+
+}  // namespace
+
+void
+katana::python::InitReductions(py::module_& m) {
+  katana::InstantiateForStandardTypes(
+      m, "ReduceSum", ReducibleFunctor<ForGAccumulator>());
+  katana::InstantiateForStandardTypes(
+      m, "ReduceMax", ReducibleFunctor<ForReduceMax>());
+  katana::InstantiateForStandardTypes(
+      m, "ReduceMin", ReducibleFunctor<ForReduceMin>());
+  ReducibleFunctor<ForReduceLogicalOr>().operator()<bool>(m, "ReduceOr");
+  ReducibleFunctor<ForReduceLogicalAnd>().operator()<bool>(m, "ReduceAnd");
```

Comment:
Is there any purpose to doing all this template metaprogramming? Don't we just need to call `katana::GReduceLogicalAnd` etc. ?

Reply:
We need to call `ReducibleFunctor` with multiple layers of template arguments so it looks complex here, but it's really just a template function call. The reason for `ForReduceLogicalOr` (and similar) is that C++14 (or whatever exactly we support) doesn't allow templates to be passed as arguments to other templates, so passing `katana::GReduceMin` directly  isn't allowed. So we need to have these wrapper structs that carry the template. Since it's required for the parametric reducers like `GReduceMin` we need to use it for `ReduceLogicalOr` as well to avoid having to duplicate `ReducibleFunctor`.

Reply:
Ok I think I kind of understand what you are trying here.

Reply:
I added a few in code comments to try to help future readers as well. I really wish templates were less stupidly designed.